File: typescript/src/com/murex/tbw/client/BooksOrder.ts
import {Book} from "@app/domain/book";
import {Invoice, PurchasedBook} from "@app/purchase";
import {MainRepository} from "../MainRepository";
import {Client, Order} from ".";

export class BooksOrder implements Order {
  private readonly client: Client;
  private readonly booksInBasket = new Map<Book, number>();

  constructor(client: Client) {
    this.client = client;
  }

  addBook(book: Book, quantity: number): void {
    const alreadySold = this.booksInBasket.get(book) || 0;
    this.booksInBasket.set(book, alreadySold + quantity);
  }

  checkOut(): Invoice {
    const invoice = new Invoice(this.client.name, this.client.country);

    this.booksInBasket.forEach((quantity: number, book: Book) => {
      invoice.addPurchasedBook(new PurchasedBook(book, quantity));
    });
    MainRepository.configuredRepository().addInvoice(invoice);
    return invoice;
  }

  getQuantityOf(book: Book): number {
    return this.booksInBasket.get(book) || 0;
  }
}

File: java/src/main/java/org/peaksys/gameoftechs/client/BooksOrder.java
package org.peaksys.gameoftechs.client;

import org.peaksys.gameoftechs.MainRepository;
import org.peaksys.gameoftechs.domain.book.Book;
import org.peaksys.gameoftechs.purchase.Invoice;
import org.peaksys.gameoftechs.purchase.PurchasedBook;

import java.util.HashMap;
import java.util.Map;

public final class BooksOrder implements Order {
    private final Map<Book, Integer> booksInBasket;
    private final Client client;

    public BooksOrder(Client client) {
        this.client = client;
        booksInBasket = new HashMap<>();
    }

    @Override
    public void addBook(Book book, int quantity) {
        booksInBasket.putIfAbsent(book, 0);
        booksInBasket.put(book, booksInBasket.get(book) + quantity);
    }

    @Override
    public Invoice checkOut() {
        Invoice invoice = new Invoice(client.name(), client.country());
        booksInBasket.forEach((book, quantity) -> invoice.addPurchasedBook(new PurchasedBook(book, quantity)));
        MainRepository.configuredRepository().addInvoice(invoice);
        return invoice;
    }

    @Override
    public Integer getQuantityOf(Book book) {
        return booksInBasket.getOrDefault(book, 0);
    }
}
